contribs/src/main/scala/io/delta/storage/BaseExternalLogStore.scala

```diff
@@ -0,0 +1,407 @@
+    } else {
+      None
+    }
+  }
+
+  /**
+   * Write file with actions under a specific path.
+   */
+  private def writeActions(fs: FileSystem,
+      path: Path,
+      actions: Iterator[String]): Long = {
+    logDebug(s"writeActions to: $path")
+    val stream = new CountingOutputStream(fs.create(path, true))
+    actions.map(_ + "\n").map(_.getBytes(UTF_8)).foreach(stream.write)
+    stream.close()
+    stream.getCount
+  }
+
+  /**
+   * Copies file within filesystem.
+   *
+   * Ensures that the `src` file is either entirely copied to the `dst` file, or not at all.
+   *
+   * @param fs  reference to [[FileSystem]]
+   * @param src path to source file
+   * @param dst path to destination file
+   */
+  private def copyFile(fs: FileSystem, src: Path, dst: Path) {
+    logDebug(s"copy file: $src -> $dst")
+    val input_stream = fs.open(src)
+    val output_stream = fs.create(dst, true)
+    try {
+      IOUtils.copy(input_stream, output_stream)
+      output_stream.close()
+    } finally {
+      input_stream.close()
+    }
+  }
+
+  /**
+   * Merge two iterators of [[FileStatus]] into a single iterator ordered by file path name.
+   * In case both iterators have [[FileStatus]]s for the same file path, keep the one from
+   * `iterWithPrecedence` and discard that from `iter`.
+   */
+  private def mergeFileIterators(
+      iter: Iterator[FileStatus],
+      iterWithPrecedence: Iterator[FileStatus]
+  ): Iterator[FileStatus] = {
+    val result = (
+        iter.map(f => (f.getPath, f)).toMap ++
+        iterWithPrecedence.map(f => (f.getPath, f))
+      )
+      .values
+      .toSeq
+      .sortBy(_.getPath.getName)
+    result.iterator
+  }
+
+  private def resolved(path: Path): (FileSystem, Path) = {
+    val fs = path.getFileSystem(getHadoopConfiguration)
+    val resolvedPath = stripUserInfo(fs.makeQualified(path))
+    (fs, resolvedPath)
+  }
+
+  private def stripUserInfo(path: Path): Path = {
+    val uri = path.toUri
+    val newUri = new URI(
+      uri.getScheme,
+      null,
+      uri.getHost,
+      uri.getPort,
+      uri.getPath,
+      uri.getQuery,
+      uri.getFragment
+    )
+    new Path(newUri)
+  }
+
+  /**
+   * Best-effort at ensuring consistency on file system according to the external cache.
+   *
+   * Tries to copy a LogEntryMetadata's temp file into the target _delta_log file, if it does not
+   * exist already, and then commit to the external cache.
+   *
+   * Will retry at most 3 times.
+   *
+   * @return the correct FileStatus from which to read the entry's data
+   */
+  private def tryFixTransactionLog(
+      fs: FileSystem,
+      entry: LogEntryMetadata,
+      failureAcceptable: Boolean): FileStatus = {
+    logDebug(s"Try to fix: ${entry.path}")
+    val completedEntry = entry.complete()
+    var fileCopied = false
+
+    for (i <- 0 until 3) {
+      try {
+        if (!fileCopied && !fs.exists(entry.path)) {
+          copyFile(fs, entry.tempPath, entry.path)
```

Comment:
copyFile overwrite should be false. so what happens if this throws a FileAlreadyExistsException?

we should not try to copy the file again. but we should still try and commit.